File: client_app/src/hooks/data-fetching/reservation.ts
import { Reservation } from "@definitions/types";
import { useRequest } from "@hooks/useRequest";
export type ReserveForm = Omit<
  Reservation,
  | "id"
  | "accountId"
  | "client"
  | "device"
  | "status"
  | "statusId"
  | "createdAt"
  | "timeSlot"
  | "dateSlot"
  | "remarks"
>;
import {
  MutationOptions,
  UseQueryOptions,
  useMutation,
  useQuery,
} from "@tanstack/react-query";

export const useNewReservation = ({
  onSuccess,
  onSettled,
  onError,
}: MutationOptions<any, unknown, ReserveForm, unknown>) => {
  const { Post } = useRequest();
  return useMutation({
    mutationFn: (form) =>
      Post(`/reservations`, form, {
        headers: {
          "content-type": "application/json",
        },
      }),
    onSuccess: onSuccess,
    onError: onError,
    onSettled: onSettled,
  });
};

export const useReservations = ({
  onSuccess,
  onError,
  onSettled,
}: UseQueryOptions<Reservation[]>) => {
  const { Get } = useRequest();
  const fetchReservations = async () => {
    try {
      const { data: response } = await Get("/reservations", {
        params: {},
      });

      const { data } = response;
      return data?.reservations ?? [];
    } catch {
      return [];
    }
  };
  return useQuery<Reservation[]>({
    queryFn: fetchReservations,
    onSuccess: onSuccess,
    onError: onError,
    queryKey: ["reservations"],
    onSettled,
  });
};

File: server/model/reservation.go
package model

import "github.com/RyanAliXII/sti-munoz-library-system/server/app/db"


type Reservation struct {
	Id string `json:"id" db:"id"`
	TimeSlotId string `json:"timeSlotId" db:"time_slot_id"`
	DateSlotId	string `json:"dateSlotId" db:"date_slot_id"`
	DeviceId string `json:"deviceId" db:"device_id"`
	AccountId string `json:"accountId" db:"account_id"`
	Client AccountJSON `json:"client" db:"client"`
	DateSlot DateSlotJSON `json:"dateSlot" db:"date_slot"`
	TimeSlot TimeSlotJSON `json:"timeSlot" db:"time_slot"`
	StatusId int `json:"statusId" db:"status_id"`
	Status string `json:"status" db:"status"`
	Device DeviceJSON `json:"device" db:"device"`
	Remarks string `json:"remarks" db:"remarks"`
	CreatedAt db.NullableTime `json:"createdAt" db:"created_at"`
}
